ompi/datatype/ompi_datatype.h

```diff
@@ -418,5 +418,94 @@ OMPI_DECLSPEC int ompi_datatype_pack_external_size( const char datarep[], int in
         }                                                               \
     }
 
+/*
+ * Sometimes it's faster to operate on a (count,datatype) pair if it's
```

Comment:
This is true in MPICH but not in OMPI, because the optimization step is supposed to automatically consolidate such cases. I am talking about the code in opal_datatype_optimize_short (more precisely opal_datatype_optimize.c around line 158). It is possible that code fails to detect the pattern, and I would love to see such an example. Can you provide an example for me to look into ?

Reply:
Let me answer my own question. The code I was pointing, identify and optimize datatypes that exhibit the pattern you mention in your comment, except when the pattern is not part of the datatype itself, but it is part of the MPI calls (and therefore has a count around a single data entry pointing to a non contiguous element).

Reply:
Yeah, if it was all contained in a single MPI datatype I think the pattern would have been detected and optimized.  It's just that a (count,dt) pair from an MPI like pack call never gets analyzed together by that code.

So in the gist testcase from the top, the two "dtbig" lines are nicely optimized, but in the two "dtsmall" cases it needs something like this datatype_consolidate structure in order to trigger the same optimization